mesonbuild/modules/freedesktop.py

```diff
@@ -0,0 +1,206 @@
+    e.text = text
+    return e
+
+
+class FreedesktopModule(ExtensionModule):
+
+    @permittedKwargs({'name', 'generic_name', 'icon', 'categories',
+                      'actions', 'install_name', 'mimetype', 'comment',
+                      'dbus'})
+    def desktop_file(self, state: 'ModuleState',
+                     args: T.Tuple[T.Union['ExecutableHolder', 'Disabler']],
+                     kwargs: T.Dict[str, T.Any]) -> ModuleReturnValue:
+        if len(args) != 1:
+            raise MesonException('Only one positinal argument allowed.')
+
+        if isinstance(args[0], Disabler):
+            return ModuleReturnValue(None, [])
+
+        exe = args[0].held_object  # type: Executable
+        if not isinstance(exe, Executable):
+            raise MesonException('First argument must be an Executable')
+
+        name = kwargs.get('name', '')  # type: str
+        if not name:
+            name = exe.name
+
+        if not exe.should_install():
+            # If we're not installing the file, don't install this
+            mlog.warning('Desktop file for', name, 'not being generated, as '
+                         'the program is not going to be installed')
+            return ModuleReturnValue(None, [])
+
+        config = FDOConfigParser()
+        values = {}
+
+        values['Name'] = name
+        values['Terminal'] = str(not exe.gui_app).lower()
+        values['Exec'] = os.path.join(exe.get_install_dir(state.environment)[0][0], exe.name)
+        values['Type'] = 'Application'
+        values['Version'] = '1.1'
+        optional_string_arg(kwargs, 'comment', values, 'Comment')
+        optional_string_arg(kwargs, 'generic_name', values, 'GenericName')
+        optional_string_arg(kwargs, 'icon', values, 'Icon')
+        optional_list_arg(kwargs, 'categories', values, 'Categories')
+        optional_list_arg(kwargs, 'mimetype', values, 'MimeType')
+
+        dbus = kwargs.get('dbus', False)
+        if not isinstance(dbus, bool):
+            raise MesonException('dbus must be a boolean.')
+        values['DBusActivatable'] = str(dbus).lower()
+
+        actions = listify(kwargs.get('actions', []))
+        if actions:
+            for a in actions:
+                if not isinstance(a, dict):
+                    raise MesonException('"actions" must be a list of dicts if defined.')
+            values['Actions'] = ';'.join(a['action'] for a in actions)
+
+        # This must happen after values['Action'], but before any action
+        # sections are added.
+        config['Desktop Entry'] = values
+
+        for action in actions:
+            if 'action' not in action:
+                raise MesonException('missing required "action" entry in action "{}"'.format(action))
+            if 'name' not in action:
+                raise MesonException('missing required "name" entry in action "{}"'.format(action))
+            avalues = {}
+            avalues['Name'] = action['name']
+
+            exec_ = listify(action['exec'])
```

Comment:
You don't validate that 'exec' is in the dict, and you could use stringlistify() to validate they are strings.